File: DataStructure/3.LinkedList/Group3/sungmin/4949.cpp
#include <iostream>
#include <stack>
#include <string>

using namespace std;

int main()
{
    string str1;

    while (1) {
        getline(cin, str1);  

        if (str1 == ".") {
            break;
        }

        stack<char> st;  

        bool isValid = true;  

        for (char ch : str1) {
            if (ch == '(' || ch == '[') {
                st.push(ch);  
            } 
            else if (ch == ')') {
                if (!st.empty() && st.top() == '(') {
                    st.pop();  
                } else {
                    isValid = false;  
                    break;
                }
            } 
            else if (ch == ']') {
                if (!st.empty() && st.top() == '[') {
                    st.pop();  
                } else {
                    isValid = false;  
                    break;
                }
            }
        }

        if (isValid && st.empty()) {
            cout << "yes" << '\n';
        } else {
            cout << "no" << '\n';
        }
    }

    return 0;
}

File: DataStructure/3.LinkedList/Group3/dooyeon/4949.py
while True:
    str1 = input().strip()
    if str1 == ".":
        break
    st = []
    valid = True
    for ch in str1:
        if ch == '(' or ch == '[':
            st.append(ch)
        elif ch == ')':
            if st and st[-1] == '(':
                st.pop()
            else:
                valid = False
                break
        elif ch == ']':
            if st and st[-1] == '[':
                st.pop()
            else:
                valid = False
                break
    print("yes" if valid and not st else "no")